src/invert/laplacexz/impls/petsc/laplacexz-petsc.cxx

```diff
@@ -386,59 +386,191 @@ void LaplaceXZpetsc::setCoefs(const Field3D &Ain, const Field3D &Bin) {
+              Acoef * J * g11 / (coords->J(x, y, z) * dx * coords->dx(x, y, z));
+          xp = val;
+          c = -val;
+        }
 
-        // Metrics on x+1/2 boundary
-        BoutReal J = 0.5*(coords->J(x,y) + coords->J(x+1,y));
-        BoutReal g11 = 0.5*(coords->g11(x,y) + coords->g11(x+1,y));
-        BoutReal dx = 0.5*(coords->dx(x,y) + coords->dx(x+1,y));
-        BoutReal Acoef = 0.5*(A(x,y,z) + A(x+1,y,z));
+        {
+          // Metrics on x-1/2 boundary
+          const BoutReal J = 0.5 * (coords->J(x, y, z) + coords->J(x - 1, y, z));
+          const BoutReal g11 = 0.5 * (coords->g11(x, y, z) + coords->g11(x - 1, y, z));
+          const BoutReal dx = 0.5 * (coords->dx(x, y, z) + coords->dx(x - 1, y, z));
+          const BoutReal Acoef = 0.5 * (A(x, y, z) + A(x - 1, y, z));
+
+          const BoutReal val =
+              Acoef * J * g11 / (coords->J(x, y, z) * dx * coords->dx(x, y, z));
+          xm = val;
+          c -= val;
+        }
 
-        BoutReal val = Acoef * J * g11 / (coords->J(x,y) * dx * coords->dx(x,y));
-        xp = val;
-        c  = -val;
+        // ZZ component
+        // Wrap around z-1 and z+1 indices
+        const int zminus = (z - 1 + (localmesh->LocalNz)) % (localmesh->LocalNz);
+        const int zplus = (z + 1) % (localmesh->LocalNz);
+
+        {
+          const BoutReal J = 0.5 * (coords->J(x, y, z) + coords->J(x, y, zplus));
+          const BoutReal g33 = 0.5 * (coords->g33(x, y, z) + coords->g33(x, y, zplus));
+          const BoutReal dz = 0.5 * (coords->dz(x, y, z) + coords->dz(x, y, zplus));
+          // Metrics on z+1/2 boundary
+          const BoutReal Acoef = 0.5 * (A(x, y, z) + A(x, y, zplus));
+
+          const BoutReal val = Acoef * J * g33 / (coords->J(x, y, z) * dz * dz);
+          zp = val;
+          c -= val;
+        }
+        {
+          // Metrics on z-1/2 boundary
+          const BoutReal J = 0.5 * (coords->J(x, y, z) + coords->J(x, y, zminus));
+          const BoutReal g33 = 0.5 * (coords->g33(x, y, z) + coords->g33(x, y, zminus));
+          const BoutReal Acoef = 0.5 * (A(x, y, z) + A(x, y, zminus));
+          const BoutReal dz = 0.5 * (coords->dz(x, y, z) + coords->dz(x, y, zminus));
+
+          const BoutReal val = Acoef * J * g33 / (coords->J(x, y, z) * dz * dz);
+          zm = val;
+          c -= val;
+        }
 
-        // Metrics on x-1/2 boundary
-        J = 0.5*(coords->J(x,y) + coords->J(x-1,y));
-        g11 = 0.5*(coords->g11(x,y) + coords->g11(x-1,y));
-        dx = 0.5*(coords->dx(x,y) + coords->dx(x-1,y));
-        Acoef = 0.5*(A(x,y,z) + A(x-1,y,z));
+        // XZ components
+
+        {
+          // x+1/2, z+1/2
+          // TODOL Should they have a z dependency?
+          const BoutReal J = 0.5 * (coords->J(x, y, z) + coords->J(x + 1, y, z));
+          const BoutReal g13 = 0.5 * (coords->g13(x, y, z) + coords->g13(x + 1, y, z));
+          const BoutReal fourdz = 2 * (coords->dz(x, y, z) + coords->dz(x + 1, y, zplus));
```

Comment:
There's a TODO here -- still outstanding?

Also `fourdz` has `zplus` in the second `dz` -- typo? I don't know the maths here, but it looks like the metric terms are at x+1/2, while `A` is at z+1/2. Actually, `coords->dx` below is at `x, y, z`, so it's definitely not obvious to me what it should be.

@bshanahan  can you double check this?

Reply:
Checking this currently... Something odd in the mixed X-Z terms, with the same matrix element being written to twice e.g:
```
MatSetValues(it.MatA, 1, &row, 1, &xzcol, &xmzp, INSERT_VALUES);
MatSetValues(it.MatA, 1, &row, 1, &xzcol, &zpxm, INSERT_VALUES);
```

Reply:
I've been through and changed this code, see PR #2204
It would be great to have an actual test to check this, but for now some ASCII art will have to suffice.

Reply:
> It would be great to have an actual test to check this, but for now some ASCII art will have to suffice.

We could test using the forward operator (i.e. Laplace_perp with the exact same discretisation as LaplaceXZ). Not sure if the forward operator exists - if not it's usually simpler to implement than the matrix for the inverse, so makes a good test. Then the test can pretty much copy-paste `test-laplacexy-fv`.

Reply:
It turns out there is a test: https://github.com/boutproject/BOUT-dev/pull/2025/files#diff-ffa989d190cd17bcd1ce0b3f23971a05d4ea8967019acb806b976f2ede1bb2a8R29

but it passes both before and after the change in #2204, I guess because `g13` is a scalar (and all other components just the identity), but to really test this we need every component to be 3D

Might be nicer to adapt a version of the `test_laplace_petsc3damg` unit test?

Reply:
This test returns the same answers with and without forcing g13 to be Field3D or leaving it as a scalar. (with and without the changes in #2204 )

Reply:
What is the forward operator? Naively, I guess something like this:

```cpp
  Field3D operator()(Field3D& f) const {
    auto result = (DDX(a * J * coord->g11 * DDX(f))   // x-x
                   + DDZ(a * J * coord->g33 * DDZ(f)) // z-z
                   + DDX(a * J * coord->g13 * DDZ(f)) // x-z
                   + DDZ(a * J * coord->g13 * DDX(f)) // z-x
                   ) / J
                  + b * f;
```

Are there coefficients on any of the terms?